packages/pds/src/labeler/base.ts

```diff
@@ -0,0 +1,78 @@
+import stream from 'stream'
+import PQueue from 'p-queue'
+import Database from '../db'
+import { BlobStore, cidForRecord } from '@atproto/repo'
+import { dedupe, getFieldsFromRecord } from './util'
+import { AtUri } from '@atproto/uri'
+import { labelerLogger as log } from '../logger'
+
+export abstract class Labeler {
+  public db: Database
+  public blobstore: BlobStore
+  public labelerDid: string
+  public processingQueue: PQueue | null // null during teardown
+  constructor(opts: {
+    db: Database
+    blobstore: BlobStore
+    labelerDid: string
+  }) {
+    this.db = opts.db
+    this.blobstore = opts.blobstore
+    this.labelerDid = opts.labelerDid
+    this.processingQueue = new PQueue()
+  }
+
+  processRecord(uri: AtUri, obj: unknown) {
+    this.processingQueue?.add(() =>
+      this.createAndStoreLabels(uri, obj).catch((err) => {
+        log.error(
+          { err, uri: uri.toString(), record: obj },
+          'failed to label record',
+        )
+      }),
+    )
+  }
+
+  async createAndStoreLabels(uri: AtUri, obj: unknown): Promise<void> {
+    const labels = await this.labelRecord(obj)
+    if (labels.length < 1) return
+    const cid = await cidForRecord(obj)
+    const rows = labels.map((value) => ({
+      sourceDid: this.labelerDid,
+      subjectUri: uri.toString(),
+      subjectCid: cid.toString(),
+      value,
+      negated: 0 as const,
+      createdAt: new Date().toISOString(),
+    }))
+
+    await this.db.db.insertInto('label').values(rows).execute()
```

Comment:
Do we need any handling for a record that has already been labeled?  E.g. maybe no-oping when trying to add a label that already exists with a `do nothing`?

Reply:
oh great call! we need an "on conflict" here